### PR TITLE
ARROW-13769: [C++] Modify BitmapOp API to fetch output validity count if requested

### PR DESCRIPTION
**Pull request by 9prady9**
Currently there are two versions of Bitmap operations in the API.

1. Returns a Buffer object
2. Updates the output pointer passed in as function argument

With this change, they can now return bitmap operation result's validity count. All bitmap ops are updated to reflect the same.

This new output argument passed to the function should be pointer to single value (only first value is used if pointer to an array is passed). If this pointer is set to `nullptr`, the output validity count isn't calculated.

### CODE REVIEW

File: cpp/src/arrow/util/bit_util_benchmark.cc [hunk 1]
```diff
@@ -159,9 +159,11 @@ static void BenchmarkAndImpl(benchmark::State& state, DoAnd&& do_and) {
 
 static void BenchmarkBitmapAnd(benchmark::State& state) {
   BenchmarkAndImpl(state, [](const internal::Bitmap(&bitmaps)[2], internal::Bitmap* out) {
+    int64_t outPopCount = 0;
     internal::BitmapAnd(bitmaps[0].buffer()->data(), bitmaps[0].offset(),
                         bitmaps[1].buffer()->data(), bitmaps[1].offset(),
-                        bitmaps[0].length(), 0, out->buffer()->mutable_data());
+                        bitmaps[0].length(), 0, out->buffer()->mutable_data(),
+                        &outPopCount);
```
Comment (cyb70289) [inline cpp/src/arrow/util/bit_util_benchmark.cc:166]: Will you append `benchmark::DoNotOptimize(outPopCount)`?

File: cpp/src/arrow/util/bitmap_ops.cc [hunk 2]
```diff
@@ -249,59 +249,92 @@ void AlignedBitmapOp(const uint8_t* left, int64_t left_offset, const uint8_t* ri
   left += left_offset / 8;
   right += right_offset / 8;
   out += out_offset / 8;
+  uint64_t outPopCount = 0;
   for (int64_t i = 0; i < nbytes; ++i) {
     out[i] = op(left[i], right[i]);
+    if (ComputeNewValidityCount) {
+      outPopCount += BitUtil::kBytePopcount[out[i]];
```
Comment (cyb70289) [inline cpp/src/arrow/util/bitmap_ops.cc:256]: This extra line may disable compiler to vectorize the loop and cause big performance regression.[1]
But from your benchmark result, there's at most 20% difference.
Would you list the test names in your benchmark result graph? BitmapAnd related benchmark is about 9G/s on my test machine (no faster than most desktop pc), looks your test result is about 400M/s.

[1] https://quick-bench.com/q/ko7aKpF3R2N8G2yqN3g5mUKgOPg

Comment (9prady9) [reply cpp/src/arrow/util/bitmap_ops.cc:256]: I have checked it with clang shouldn't be different with gcc/msvc. Here is the godbolt snippet with vectorization flags showing same type of vectorization happening with pop-count and without. https://godbolt.org/z/bGhdGEjr6

Yeah, M/s is my mistake. Those are numbers from debug build I did, I forgot to switch to release. With release, the  numbers are much higher. I shall update the description with new numbers soon and with benchmark labels.

Comment (cyb70289) [reply cpp/src/arrow/util/bitmap_ops.cc:256]: > Here is the godbolt snippet with vectorization flags showing same type of vectorization happening with pop-count and without. https://godbolt.org/z/bGhdGEjr6 But I will check with gcc/msvc too

In your test code, line 34~43.
The problem is that `v` is calculated but never used. So compiler will simply delete all the code calculating it.
Change the function to let it return `v`, the generated code will be totally different.

```cpp
void WithPopcount(const uint8_t* a, const uint8_t* b, const int N) {
  std::vector<uint8_t> c;
  c.reserve(N);

  unsigned int v = 0;
  for (int i = 0; i < N; ++i) {
    c[i] = a[i] & b[i];
    v += table[c[i]];
  }
}
```

> Yeah, M/s is my mistake. Those are numbers from debug build I did, I forgot to switch to release. With release, the numbers are much higher. I shall update the description with new numbers soon and with benchmark labels.

No worry, it happens.

Comment (9prady9) [reply cpp/src/arrow/util/bitmap_ops.cc:256]: Upon quick check, similar isn't observed on gcc and msvc. Will push updates once they are ready. Thank you for the feedback!

Comment (9prady9) [reply cpp/src/arrow/util/bitmap_ops.cc:256]: > The problem is that v is calculated but never used. So compiler will simply delete all the code calculating it.
> Change the function to let it return v, the generated code will be totally different.

Makes sense. After playing with it for few minutes, saving result to register and then using that for other two calculations is what letting compiler vectorize the loop on all three compilers consistently.

- clang - https://godbolt.org/z/sKvodGzbr
- gcc - https://godbolt.org/z/Kdc38P7fx
- msvc - https://godbolt.org/z/n1so9PbG3

In any case, eventually I dropped the above approach. 

I had a version of CountSetBits earlier that was doing popcount on 64-bit words. But I removed that in favour of lookup during the bitmap op assuming lookup might be faster. It turned out just doing popcount finally on the whole array is more efficient than clubbing the pop-count with bitmapOp .

Please check the updated description for new numbers

**Comment by 9prady9** [comment]
The difference in performance is mostly with bitmap op code path that is optimized for aligned access.

File: cpp/src/arrow/util/bitmap_ops.cc [hunk 3]
```diff
@@ -237,71 +237,105 @@ bool OptionalBitmapEquals(const std::shared_ptr<Buffer>& left, int64_t left_offs
...
+    outPopCount = CountSetBits(outFront, out_offset, length);
+  }
+  return outPopCount;
 }
 
-template <template <typename> class BitOp>
-void UnalignedBitmapOp(const uint8_t* left, int64_t left_offset, const uint8_t* right,
-                       int64_t right_offset, uint8_t* out, int64_t out_offset,
-                       int64_t length) {
+template <template <typename> class BitOp, bool ComputeNewValidityCount>
```
Comment (cyb70289) [inline cpp/src/arrow/util/bitmap_ops.cc:263]: From your test result, looks pop count calculation doesn't impact performance.
If that's the case, I think we should remove this new template argument and always calc and return pop count in this function. Caller can simply ignore the return value if it don't need it.

File: cpp/src/arrow/util/bitmap_ops.cc [hunk 4]
```diff
@@ -237,71 +237,105 @@ bool OptionalBitmapEquals(const std::shared_ptr<Buffer>& left, int64_t left_offs
...
+    writer.PutNextWord(outWord);
   }
   auto nbytes = left_reader.trailing_bytes();
   while (nbytes--) {
     int left_valid_bits, right_valid_bits;
     uint8_t left_byte = left_reader.NextTrailingByte(left_valid_bits);
     uint8_t right_byte = right_reader.NextTrailingByte(right_valid_bits);
     DCHECK_EQ(left_valid_bits, right_valid_bits);
-    writer.PutNextTrailingByte(op_byte(left_byte, right_byte), left_valid_bits);
+    auto outByte = op_byte(left_byte, right_byte);
```
Comment (cyb70289) [inline cpp/src/arrow/util/bitmap_ops.cc:289]: use snake case `out_byte`

File: cpp/src/arrow/util/bitmap_ops.cc [hunk 5]
```diff
@@ -311,39 +345,49 @@ Result<std::shared_ptr<Buffer>> BitmapAnd(MemoryPool* pool, const uint8_t* left,
                                           int64_t left_offset, const uint8_t* right,
                                           int64_t right_offset, int64_t length,
                                           int64_t out_offset) {
+  // TODO(pradeep) Figure out the return type for this version of BitmapOps
+  int64_t out_pop_count = 0;
   return BitmapOp<std::bit_and>(pool, left, left_offset, right, right_offset, length,
-                                out_offset);
+                                out_offset, &out_pop_count);
```
Comment (cyb70289) [inline cpp/src/arrow/util/bitmap_ops.cc:351]: I don't understand why this `out_pop_count`. It looks useless.

File: cpp/src/arrow/array/array_nested.cc [hunk 6]
```diff
@@ -680,7 +680,7 @@ Result<std::shared_ptr<Array>> SparseUnionArray::GetFlattenedField(
   if (child_null_bitmap) {
     BitmapAnd(flattened_null_bitmap->data(), child_offset, child_null_bitmap->data(),
               child_offset, child_data->length, child_offset,
-              flattened_null_bitmap->mutable_data());
+              flattened_null_bitmap->mutable_data(), /*out pop count*/ nullptr);
```
Comment (cyb70289) [inline cpp/src/arrow/array/array_nested.cc:683]: Make`nullptr` the default value to this new argument to leave current code untouched if they don't care about pop count.

File: cpp/src/arrow/util/bitmap_ops.h [hunk 7]
```diff
@@ -124,7 +124,8 @@ Result<std::shared_ptr<Buffer>> BitmapAnd(MemoryPool* pool, const uint8_t* left,
 /// the results in out starting at the given bit-offset.
 ARROW_EXPORT
 void BitmapAnd(const uint8_t* left, int64_t left_offset, const uint8_t* right,
-               int64_t right_offset, int64_t length, int64_t out_offset, uint8_t* out);
+               int64_t right_offset, int64_t length, int64_t out_offset, uint8_t* out,
+               int64_t* const out_pop_count);
```
Comment (cyb70289) [inline cpp/src/arrow/util/bitmap_ops.h:128]: remove `const`, and make `nullptr` the default value as commented.

File: cpp/src/arrow/util/bit_util_benchmark.cc [hunk 8]
```diff
@@ -159,9 +159,12 @@ static void BenchmarkAndImpl(benchmark::State& state, DoAnd&& do_and) {
 
 static void BenchmarkBitmapAnd(benchmark::State& state) {
   BenchmarkAndImpl(state, [](const internal::Bitmap(&bitmaps)[2], internal::Bitmap* out) {
+    int64_t outPopCount = 0;
```
Comment (cyb70289) [inline cpp/src/arrow/util/bit_util_benchmark.cc:162]: use snake case

File: cpp/src/arrow/util/bit_util_benchmark.cc [hunk 9]
```diff
@@ -159,9 +159,12 @@ static void BenchmarkAndImpl(benchmark::State& state, DoAnd&& do_and) {
...
     internal::BitmapAnd(bitmaps[0].buffer()->data(), bitmaps[0].offset(),
                         bitmaps[1].buffer()->data(), bitmaps[1].offset(),
-                        bitmaps[0].length(), 0, out->buffer()->mutable_data());
+                        bitmaps[0].length(), 0, out->buffer()->mutable_data(),
+                        &outPopCount);
+    benchmark::DoNotOptimize(outPopCount);
```
Comment (cyb70289) [inline cpp/src/arrow/util/bit_util_benchmark.cc:167]: Augment current benchmark to add PopCount tests. Popcount is not used in many cases.

File: cpp/src/arrow/util/bit_util_test.cc [hunk 10]
```diff
@@ -1357,10 +1372,15 @@ class BitmapOp : public TestBase {
 
           // Clear out buffer and try non-allocating version
           std::memset(out->mutable_data(), 0, out->size());
+          int64_t expectedPopCount = 0;
```
Comment (cyb70289) [inline cpp/src/arrow/util/bit_util_test.cc:1375]: I think it should be simply `pop_count`, not expected.

File: cpp/src/arrow/util/bit_util_test.cc [hunk 11]
```diff
@@ -1357,10 +1372,15 @@ class BitmapOp : public TestBase {
...
           ASSERT_OK(op.Call(left->mutable_data(), left_offset, right->mutable_data(),
-                            right_offset, length, out_offset, out->mutable_data()));
+                            right_offset, length, out_offset, out->mutable_data(),
+                            &expectedPopCount));
           reader = internal::BitmapReader(out->mutable_data(), out_offset, length);
           ASSERT_READER_VALUES(reader, result_bits);
+
+          auto refPopCount = SlowCountBits(out->mutable_data(), out_offset, length);
```
Comment (cyb70289) [inline cpp/src/arrow/util/bit_util_test.cc:1382]: We usaully call this `expected_pop_count` in arrow unit tests.

**Review by cyb70289 — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/array/array_nested.cc (hunk 6 above)
Comment (9prady9) [reply cpp/src/arrow/array/array_nested.cc:683]: I did think about that, but wasn't sure if default arguments are used in exported functions, especially in the functions that had pointers and basic types compared to bitmap functions that returned class objects.

Is there any C API in arrow where this would apply ? such as API's that are wrapped around from other languages like Rust ?

File: cpp/src/arrow/util/bitmap_ops.cc (hunk 5 above)
Comment (9prady9) [reply cpp/src/arrow/util/bitmap_ops.cc:351]: Wanted to ask the arrow devs if the functions returning class objects also need to be modified to return pop count. It is an incomplete change ( I though I left this PR in draft mode, hm). Please share your thoughts - if you don't want to change these versions of bitmap API calls or otherwise. Hence left the TODO line above this function call https://github.com/apache/arrow/pull/11674/files#diff-1162624f731f4758d9b900e493a4551a266ea06d081f7ade43a21e333a63d8c0R348

File: cpp/src/arrow/util/bitmap_ops.cc (hunk 3 above)
Comment (9prady9) [reply cpp/src/arrow/util/bitmap_ops.cc:263]: No, there is an performance difference, quite good I would think - for the aligned code-path. Hence, I added the template bool expecting the compiler to optimize out the validity count calculation when it is requested.

File: cpp/src/arrow/util/bitmap_ops.h (hunk 7 above)
Comment (9prady9) [reply cpp/src/arrow/util/bitmap_ops.h:128]: I added this `const` qualifier to indicate pointer wouldn't change.

File: cpp/src/arrow/util/bit_util_benchmark.cc (hunk 9 above)
Comment (9prady9) [reply cpp/src/arrow/util/bit_util_benchmark.cc:167]: I didn't quite understand, do you mean test for pop count results in bitmap ops tests ? I already added that portion.

File: cpp/src/arrow/util/bitmap_ops.cc (hunk 3 above)
Comment (cyb70289) [reply cpp/src/arrow/util/bitmap_ops.cc:257]: But your benchmark graph shows no difference from unaligned op, and this function is only for unaligned op, no?

File: cpp/src/arrow/util/bit_util_benchmark.cc (hunk 9 above)
Comment (cyb70289) [reply cpp/src/arrow/util/bit_util_benchmark.cc]: Sorry I didn't say it clearly. I mean leave current benchmark untouched, and add new benchmarks for `bitmapand+popcount` cases.

File: cpp/src/arrow/util/bitmap_ops.h (hunk 7 above)
Comment (cyb70289) [reply cpp/src/arrow/util/bitmap_ops.h]: `const` parameter may be useful in definition, but putting it in declaration is a bit weird IMO.
Another reason is just to be consistent with existing parameters.

File: cpp/src/arrow/array/array_nested.cc (hunk 6 above)
Comment (cyb70289) [reply cpp/src/arrow/array/array_nested.cc]: Good questions. AFAIK, arrow c++ is called by arrow c/python/r implementation. Not sure if an exported c++ function with default parameter is problematic.
@lidavidm @kou

File: cpp/src/arrow/util/bitmap_ops.cc (hunk 5 above)
Comment (cyb70289) [reply cpp/src/arrow/util/bitmap_ops.cc]: I think it's better to return popcount for both versions.

File: cpp/src/arrow/array/array_nested.cc (hunk 6 above)
Comment (lidavidm) [reply cpp/src/arrow/array/array_nested.cc]: It should not be a problem. One, we make use of default parameters quite extensively already, and if we need a binding, we can handle it appropriately (if needed with an "extern C" though I don't think we need resort to that). Two, this function is mostly used internally and is unlikely to be used by a binding (and it's in the ::internal namespace anyways).

Comment (kou) [reply cpp/src/arrow/array/array_nested.cc]: I'm not sure that the problem case correctly but I think that there are no problem with default parameter.
(Is this ABI compatibility problem?)
Because we increase major version for each release.

File: cpp/src/arrow/util/bitmap_ops.cc (hunk 5 above)
Comment (9prady9) [reply cpp/src/arrow/util/bitmap_ops.cc]: In the function that is returning output value, should I make it pair of int64_t and Array ? what is usually done in arrow ?

File: cpp/src/arrow/util/bitmap_ops.h (hunk 7 above)
Comment (9prady9) [reply cpp/src/arrow/util/bitmap_ops.h]: Sure

File: cpp/src/arrow/util/bit_util_benchmark.cc (hunk 9 above)
Comment (9prady9) [reply cpp/src/arrow/util/bit_util_benchmark.cc]: Done, will push all updates in one go.

File: cpp/src/arrow/util/bitmap_ops.cc (hunk 5 above)
Comment (cyb70289) [reply cpp/src/arrow/util/bitmap_ops.cc]: I think we can make the api same as the void func. Adding an output int64 argument with default value nullptr is okay.

File: cpp/src/arrow/util/bitmap_ops.cc (hunk 3 above)
Comment (9prady9) [reply cpp/src/arrow/util/bitmap_ops.cc:257]: Yes that is true. Was hoping this might shave off any additional performance in rare cases at the cost of one additional template argument. I can remove it.

File: cpp/src/arrow/util/bitmap_ops.cc (hunk 5 above)
Comment (9prady9) [reply cpp/src/arrow/util/bitmap_ops.cc]: okay, it looks odd though :). So about `nullptr`, the linter is throwing up about it, May I use `NULL` ?

Comment (cyb70289) [reply cpp/src/arrow/util/bitmap_ops.cc]: `NULLPTR` is okay.

File: cpp/src/arrow/util/bitmap_ops.cc [hunk 12]
```diff
@@ -255,53 +255,71 @@ void AlignedBitmapOp(const uint8_t* left, int64_t left_offset, const uint8_t* ri
...
   } else {
     // Unaligned
-    UnalignedBitmapOp<BitOp>(left, left_offset, right, right_offset, dest, out_offset,
-                             length);
+    if (out_validity_count) {
+      *out_validity_count = UnalignedBitmapOp<BitOp>(
+          left, left_offset, right, right_offset, dest, out_offset, length);
+    } else {
+      UnalignedBitmapOp<BitOp>(left, left_offset, right, right_offset, dest, out_offset,
+                               length);
```
Comment (edponce) [inline cpp/src/arrow/util/bitmap_ops.cc:308]: I suggest to have `AlignedBitmapOp` and `UnalignedBitmapOp` have the same API, either both return the popcount or neither does it.
Based on your performance measures, you can simply call `CountSetBits` at the end of this function when validity count is "requested".

**Review by edponce — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/util/bitmap_ops.cc (hunk 12 above)
Comment (9prady9) [reply cpp/src/arrow/util/bitmap_ops.cc:308]: yes, that is true. However, when I do that there is a good 10-12% performance drop for unaligned bitmap ops also.

**Comment by 9prady9** [comment]
Benchmark prior to change:
```
--------------------------------------------------------------------------------------
Benchmark                            Time             CPU   Iterations UserCounters...
--------------------------------------------------------------------------------------
BenchmarkBitmapAnd/32768/0        1660 ns         1659 ns       420530 bytes_per_second=18.3977G/s
BenchmarkBitmapAnd/131072/0       6532 ns         6526 ns       104671 bytes_per_second=18.7054G/s
BenchmarkBitmapAnd/32768/1        3546 ns         3545 ns       197074 bytes_per_second=8.60981G/s
BenchmarkBitmapAnd/131072/1      13851 ns        13834 ns        49892 bytes_per_second=8.8239G/s
BenchmarkBitmapAnd/32768/2        3544 ns         3542 ns       197488 bytes_per_second=8.61495G/s
BenchmarkBitmapAnd/131072/2      13947 ns        13931 ns        49699 bytes_per_second=8.7624G/s
```

Benchmark after change:
```
BenchmarkBitmapAnd/32768/0                    1651 ns         1650 ns       417738 bytes_per_second=18.4979G/s
BenchmarkBitmapAnd/131072/0                   6482 ns         6473 ns        98042 bytes_per_second=18.8595G/s
BenchmarkBitmapAnd/32768/1                    3745 ns         3742 ns       186783 bytes_per_second=8.15504G/s
BenchmarkBitmapAnd/131072/1                  14781 ns        14762 ns        46967 bytes_per_second=8.26913G/s
BenchmarkBitmapAnd/32768/2                    3750 ns         3747 ns       186801 bytes_per_second=8.14497G/s
BenchmarkBitmapAnd/131072/2                  14696 ns        14678 ns        47217 bytes_per_second=8.31677G/s
BenchmarkBitmapAndWithPopCount/32768/0        2339 ns         2337 ns       299171 bytes_per_second=13.0583G/s
BenchmarkBitmapAndWithPopCount/131072/0       9293 ns         9280 ns        73666 bytes_per_second=13.1537G/s
BenchmarkBitmapAndWithPopCount/32768/1        3760 ns         3756 ns       186621 bytes_per_second=8.12568G/s
BenchmarkBitmapAndWithPopCount/131072/1      14724 ns        14702 ns        47227 bytes_per_second=8.30286G/s
BenchmarkBitmapAndWithPopCount/32768/2        3755 ns         3751 ns       186443 bytes_per_second=8.13566G/s
BenchmarkBitmapAndWithPopCount/131072/2      14698 ns        14680 ns        47458 bytes_per_second=8.31524G/s
```
please note that I have rebased and pushed again, no new changes.

**Comment by edponce** [comment]
@9prady9 In the benchmark results above, why the performance drop occurs for the first pair of runs (from 18 to 13 G/s) but not for the other runs?

**Comment by 9prady9** [comment]
> @9prady9 In the benchmark results above, why the performance drop occurs for the first pair of `BenchmarkBitmapAndWithPopCount` runs (from 18 to 13 G/s) but not for the other runs?

Aligned bitmap ops does have performance drop irrespective of whether we compute the validity count inside `AlignedBitmapOp` function or outside it by calling `CountSetBits`. When done inside the `AlignedBitmapOp` function, the drop is significantly higher.

File: cpp/src/arrow/util/bitmap_ops.h [hunk 13]
```diff
@@ -117,14 +117,16 @@ ARROW_EXPORT
 Result<std::shared_ptr<Buffer>> BitmapAnd(MemoryPool* pool, const uint8_t* left,
                                           int64_t left_offset, const uint8_t* right,
                                           int64_t right_offset, int64_t length,
-                                          int64_t out_offset);
+                                          int64_t out_offset,
+                                          int64_t* out_validity_count = nullptr);
```
Comment (edponce) [inline cpp/src/arrow/util/bitmap_ops.h:121]: Use macro `NULLPTR` instead of C++ `nullptr` because these are public APIs.

File: cpp/src/arrow/util/bitmap_ops.cc [hunk 14]
```diff
@@ -255,53 +255,71 @@ void AlignedBitmapOp(const uint8_t* left, int64_t left_offset, const uint8_t* ri
...
   }
 }
 
 template <template <typename> class BitOp>
 Result<std::shared_ptr<Buffer>> BitmapOp(MemoryPool* pool, const uint8_t* left,
                                          int64_t left_offset, const uint8_t* right,
                                          int64_t right_offset, int64_t length,
-                                         int64_t out_offset) {
+                                         int64_t out_offset,
+                                         int64_t* out_validity_count = nullptr) {
```
Comment (edponce) [inline cpp/src/arrow/util/bitmap_ops.cc:318]: `nullptr` --> `NULLPTR`

**Review by edponce — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/util/bitmap_ops.cc [hunk 15]
```diff
@@ -255,53 +255,71 @@ void AlignedBitmapOp(const uint8_t* left, int64_t left_offset, const uint8_t* ri
...
 
 template <template <typename> class BitOp>
 void BitmapOp(const uint8_t* left, int64_t left_offset, const uint8_t* right,
-              int64_t right_offset, int64_t length, int64_t out_offset, uint8_t* dest) {
+              int64_t right_offset, int64_t length, int64_t out_offset, uint8_t* dest,
+              int64_t* out_validity_count) {
   if ((out_offset % 8 == left_offset % 8) && (out_offset % 8 == right_offset % 8)) {
     // Fast case: can use bytewise AND
+    // return AlignedBitmapOp<BitOp>(left, left_offset, right, right_offset, dest,
+    // out_offset,
```
Comment (lidavidm) [inline cpp/src/arrow/util/bitmap_ops.cc:295]: nit: commented code?

**Review by lidavidm — COMMENTED** [review]
The implementation looks fine, but I wonder if this is worth it given the performance drop? At least, we would need to show that we can make use of this in some kernels to make up for it. (That said, I suppose in the context of a kernel, the overhead here matters less.)

File: cpp/src/arrow/util/bitmap_ops.cc (hunk 15 above)
Comment (9prady9) [reply cpp/src/arrow/util/bitmap_ops.cc:295]: It was there from earlier, I wasn't sure if I should remove it. Hence, left it as it is.

Comment (lidavidm) [reply cpp/src/arrow/util/bitmap_ops.cc:295]: I don't see it in the original file: https://github.com/apache/arrow/blob/23d19ce754faa32878bd5e48f56e7510d34f15c6/cpp/src/arrow/util/bitmap_ops.cc

Comment (9prady9) [reply cpp/src/arrow/util/bitmap_ops.cc]: hm, split memory. May be I only added it then. Sorry, removed it.

File: cpp/src/arrow/util/bitmap_ops.cc [hunk 16]
```diff
@@ -255,53 +255,68 @@ void AlignedBitmapOp(const uint8_t* left, int64_t left_offset, const uint8_t* ri
...
   }
 }
 
 template <template <typename> class BitOp>
 Result<std::shared_ptr<Buffer>> BitmapOp(MemoryPool* pool, const uint8_t* left,
                                          int64_t left_offset, const uint8_t* right,
                                          int64_t right_offset, int64_t length,
-                                         int64_t out_offset) {
+                                         int64_t out_offset,
+                                         int64_t* out_validity_count = NULLPTR) {
```
Comment (kou) [inline cpp/src/arrow/util/bitmap_ops.cc:315]: Do we need ` = NULLPTR` here?

File: cpp/src/arrow/util/bit_util_test.cc [hunk 17]
```diff
@@ -2247,7 +2263,7 @@ TEST(Bitmap, VisitWordsAnd) {
 
         BitmapAnd(bitmaps[0].buffer()->data(), bitmaps[0].offset(),
                   bitmaps[1].buffer()->data(), bitmaps[1].offset(), bitmaps[0].length(),
-                  0, expected_buffer->mutable_data());
+                  0, expected_buffer->mutable_data(), /*out pop count*/ nullptr);
```
Comment (kou) [inline cpp/src/arrow/util/bit_util_test.cc:2266]: It seems that we can revert this change.

**Review by kou — APPROVED** [review]
+1

**Comment by 9prady9** [comment]
I have forgot to modify the test of bitmap functions that returned buffer objects. Now I added that as well. @kou thanks for the catch on nullptr, which led me to this check.

**Comment by 9prady9** [comment]
I am unable to reproduce CI failure of `AMD 64 Windows 2019` job locally(Linux though). Any hints why this could be happening with unrelated test ?

**Comment by pitrou** [comment]
@9prady9 You can disregard the `arrow-threading-utility-test` failure, it looks entirely unrelated.

**Comment by 9prady9** [comment]
> That said, I agree with @lidavidm : this may not be worth it. I experimented locally and it's not easy to do this without making performance suffer. My original intuition was that the CPU could easily parallelize popcounts with the load/stores and logical operations, but that seems to depend on compiler optimizations quite a bit.

True, unless pop count on a byte is close to a noop (via some special instruction), popcount along with bitmap will incur additional cost.

Having said that, since validity/null count on output is only provided on explicit user request, wouldn't it be nice to have the extra feature ? one less API call for user to get pop count on output. I assuming these functions are user facing(as they are exported), hence my comment.

**Comment by lidavidm** [comment]
> I assuming these functions are user facing(as they are exported), hence my comment.

Ah, exporting vs not exporting is just whether the symbols are exposed in the resulting libraries (assuming you mean the ARROW_EXPORT macro); these functions live in the arrow::internal namespace and aren't meant to be used by end users (or at least, we don't really consider them part of the API/they're use-at-your-own-risk). These are all used internally to implement kernels and other things.

**Comment by 9prady9** [comment]
> > I assuming these functions are user facing(as they are exported), hence my comment.
> 
> Ah, exporting vs not exporting is just whether the symbols are exposed in the resulting libraries (assuming you mean the ARROW_EXPORT macro); these functions live in the arrow::internal namespace and aren't meant to be used by end users (or at least, we don't really consider them part of the API/they're use-at-your-own-risk). These are all used internally to implement kernels and other things.

Ah, in that case, it seems to me like there is no need for this change. Any internal use case that requires pop count on output can call the pop-count helper after output is computed right ? Especially, given (as per benchmark results we have seen so far) that co-computing output pop-count isn't giving any performance benefits.

**Comment by cyb70289** [comment]
> May I close this PR and leave a corresponding comment on JIRA issue ?

I'm okay to close this PR.
Thanks @9prady9 for investigating this problem. This is very useful.
